src/palimpzest/execution/execution_engine.py

```diff
@@ -0,0 +1,333 @@
+                      "num_samples": num_samples,
+                      "max_workers": self.num_workers_per_thread}
+                      for plan in plans],
+                )
+            )
+        # results = list(map(lambda x: self.execute_plan(**x),
+        #         [{"plan": plan,
+        #             "num_samples": num_samples,
+        #             "max_workers": 1}
+        #             for plan in plans],
+        #     )
+        # )
+        # split results into per-plan records and plan stats
+        all_records, all_plan_stats = zip(*results)
+
+        # process results to get sample execution data and sentinel plan stats
+        all_sample_execution_data, return_records = [], []
+        for records, plan_stats, plan in zip(all_records, all_plan_stats, plans):
+            # aggregate sentinel est. data
+            for operator_stats in plan_stats.operator_stats.values():
+                all_sample_execution_data.extend(operator_stats.record_op_stats_lst)
+
+            # if this is the max quality plan for this set of plans, return its results for these records 
+            if plan.plan_id == max_quality_plan_id:
+                return_records = records
+
+        return all_sample_execution_data, return_records, plan_stats
+
+
+    def execute_optimal_strategy(
+            self,
+            dataset: Set,
+            optimizer: Optimizer,
+            execution_data: List[RecordOpStats]=[],
+        ) -> Tuple[List[DataRecord], List[PlanStats]]:
+        # get the optimal plan according to the optimizer
+        plans = optimizer.optimize(dataset)
+        final_plan = plans[0]
+
+        # execute the plan
+        records, plan_stats = self.execute_plan(
+            plan=final_plan,
+            max_workers=self.max_workers,
+        )
+
+        # return the output records and plan stats
+        return records, [plan_stats]
+
+
+    def execute_confidence_interval_strategy(
+            self,
+            dataset: Set,
+            optimizer: Optimizer,
+            execution_data: List[RecordOpStats]=[],
+        ) -> Tuple[List[DataRecord], List[PlanStats]]:
+        # initialize output records and plan stats
+        records, plan_stats = [], []
+
+        # get the initial set of optimal plans according to the optimizer
+        plans = optimizer.optimize(dataset)
+        still_processing = True
+        while len(plans) > 1 and still_processing:
+            # identify the plan with the highest quality in the set
+            max_quality_plan_id = self.get_max_quality_plan_id(plans)
+
+            # execute the set of plans for a fixed number of samples
+            new_execution_data, new_records, new_plan_stats = self.execute_plans(list(plans), max_quality_plan_id, self.num_samples)
+            records.extend(new_records)
+            plan_stats.extend(new_plan_stats)
+
+            # set still processing to False if there are no new records
+            if new_records == []:
```

Comment:
Is there a safeguard against having empty new_records e.g., due to not passing a filter vs. being the last batch of execution?

Reply:
Good catch! Updating this logic to check the datasource length against the scan idx